### DISCUSSION TITLE
Logging incoming (GraphQL) requests to table

### DISCUSSION THREAD

**Discussion by FelixZY** [supabase/supabase#21520] [Questions]
I'm not really sure where to ask this but here goes:

I'm working to set up an open API. As this means I have very little control over the incoming queries, I would be most interested in logging incoming requests to a table for statistical purposes. This is true for REST calls but maybe even more relevant for GraphQL requests, where I would definitely want to know what queries are the most common, their execution time and maybe even the actual SQL query it was translated into. Such data would allow me to optimise for the most common requests.

I'm thinking something like:

| id INT | hash TEXT | query TEXT | sql_translation TEXT | execution_time INT | timestamp TIMESTAMPTZ |
|--|--|--|--|--|--|
| 1 | sha(query) | query abc { ... } | 'SELECT * FROM ...' | 27 | 2024-02-... |

I think it would be really neat for such log-to-table functionality to be part of [pg_graphql](https://github.com/supabase/pg_graphql). Logging to a table would also allow me to set up triggers for custom behavior. Perhaps triggers could also allow me to run custom code to fail certain requests (e.g. by failing the insert) 🤔 - loose ideas at this point.

I'm also playing with the idea to allow (anonymous) users to look up their "usage" and query performance using a special identifier, such as a special request header, UA, query id, query hash or something else. Having request logs in a table would be perfect for this.

I'm aware that the supabase platform has separate logging functionality. However, my current understanding is that it does not match my use case very well - logging to a table would be much preferred.

---

TL;DR

I'm developing an open API with anonymous end users. I'm wondering:
1. if there is a way to log incoming requests to a table for trigger integration and further processing?
2. If there is a way to log incoming GraphQL requests *specifically*, preferably as part of [pg_graphql(https://github.com/supabase/pg_graphql)?
  * If this functionality does not exist, please consider this a feature request.
3. If supabase offers a way to inspect the performance of GraphQL requests?
  * I'm outlining some basic information that would be interesting to find and use for aggregation. GraphQL is a fairly recent addition to my skillset and this outline is probably not complete as a result.
4. How to best expose my logs and their aggregates as part of my api, given current platform capabilities?
